### PR TITLE
refactor(common): decorators for credential caching

### PR DESCRIPTION
**Pull request by coryan**
Remove the `AuthorizationHeader()` function from all the `oauth2_internal::*Credential` classes. The new function can be used with a decorator to implement caching in a single place. The new implementations are much simpler, and the tests can be simplified too.

Part of the work for #10316

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10362)

### CODE REVIEW

File: google/cloud/internal/oauth2_compute_engine_credentials.cc [hunk 1]
```diff
@@ -105,16 +101,26 @@ ComputeEngineCredentials::ComputeEngineCredentials(
   }
 }
 
-StatusOr<std::pair<std::string, std::string>>
-ComputeEngineCredentials::AuthorizationHeader() {
-  std::unique_lock<std::mutex> lock(mu_);
-  return refreshing_creds_.AuthorizationHeader([this] { return Refresh(); });
+StatusOr<internal::AccessToken> ComputeEngineCredentials::GetToken(
+    std::chrono::system_clock::time_point tp) {
+  // Ignore failures fetching the account metadata, we can still get a token
```
Comment (devbww) [inline google/cloud/internal/oauth2_compute_engine_credentials.cc:106]: s/, we/. We/

File: google/cloud/internal/oauth2_service_account_credentials_test.cc [hunk 2]
```diff
@@ -873,6 +686,7 @@ TEST(ServiceAccountCredentialsTest, ParseServiceAccountRefreshResponse) {
   ScopedEnvironment disable_self_signed_jwt(
       "GOOGLE_CLOUD_CPP_EXPERIMENTAL_DISABLE_SELF_SIGNED_JWT", "1");
 
+  auto const expires_in = 1000;
```
Comment (devbww) [inline google/cloud/internal/oauth2_service_account_credentials_test.cc:689]: Make it a `std::chrono` duration?

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/oauth2_compute_engine_credentials.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/oauth2_compute_engine_credentials.cc]: Fixed.

File: google/cloud/internal/oauth2_service_account_credentials_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/oauth2_service_account_credentials_test.cc]: Done, though I have been avoiding too many test cleanups in these changes.

**Pull request merged**